### ISSUE TITLE
Install develop on Cheyenne - cdo error

### ISSUE THREAD

**Issue by climbfuji** [NOAA-EMC/hpc-stack#200]
This is the first of several issues I am going to create wrt installing hpc-stack develop on Cheyenne.

We had `cdo` disabled in `config/stack_ncar.yaml`, because this has failed in the past (for both Intel 19.1.0 and GNU-9.1.0), this time trying with GNU-10.1.0:
```
...
libtool: link: mpicc -fPIC -o cdi cdi.o printinfo.o  -L/glade/work/heinzell/fv3/hpc-stack-cheyenne-gnu-10.1.0/gnu-10.1.0/mpt-2.22/netcdf/4.7.4/lib -L/glade/work/heinzell/fv3/hpc-stack-cheyenne-gnu-10.1.0/gnu-10.1.0/mpt-2.22/hdf5/1.10.6/lib -L/glade/work/heinzell/fv3/hpc-stack-cheyenne-gnu-10.1.0/gnu-10.1.0/zlib/1.2.11/lib /glade/work/heinzell/fv3/hpc-stack-cheyenne-gnu-10.1.0/gnu-10.1.0/mpt-2.22/netcdf/4.7.4/lib/libnetcdf.a ../src/.libs/libcdi.a /glade/work/heinzell/fv3/hpc-stack-cheyenne-gnu-10.1.0/gnu-10.1.0/mpt-2.22/hdf5/1.10.6/lib/libhdf5_hl.a /glade/work/heinzell/fv3/hpc-stack-cheyenne-gnu-10.1.0/gnu-10.1.0/mpt-2.22/hdf5/1.10.6/lib/libhdf5.a -lz -ldl -lm -luuid
/usr/bin/ld: /glade/work/heinzell/fv3/hpc-stack-cheyenne-gnu-10.1.0/gnu-10.1.0/mpt-2.22/netcdf/4.7.4/lib/libnetcdf.a(libnchdf5_la-hdf5internal.o): in function `nc4_reform_coord_var':
hdf5internal.c:(.text+0xa26): undefined reference to `H5DSdetach_scale'
/usr/bin/ld: /glade/work/heinzell/fv3/hpc-stack-cheyenne-gnu-10.1.0/gnu-10.1.0/mpt-2.22/netcdf/4.7.4/lib/libnetcdf.a(libnchdf5_la-hdf5open.o): in function `get_type_info2':
hdf5open.c:(.text+0x163): undefined reference to `H5Tget_native_type'
/usr/bin/ld: /glade/work/heinzell/fv3/hpc-stack-cheyenne-gnu-10.1.0/gnu-10.1.0/mpt-2.22/netcdf/4.7.4/lib/libnetcdf.a(libnchdf5_la-hdf5open.o): in function `read_coord_dimids':
hdf5open.c:(.text+0x546): undefined reference to `H5Aopen_name'
/usr/bin/ld: /glade/work/heinzell/fv3/hpc-stack-cheyenne-gnu-10.1.0/gnu-10.1.0/mpt-2.22/netcdf/4.7.4/lib/libnetcdf.a(libnchdf5_la-hdf5open.o): in function `nc4_open_file':
hdf5open.c:(.text+0x1595): undefined reference to `H5Pset_fapl_core'
/usr/bin/ld: /glade/work/heinzell/fv3/hpc-stack-cheyenne-gnu-10.1.0/gnu-10.1.0/mpt-2.22/netcdf/4.7.4/lib/libnetcdf.a(libnchdf5_la-hdf5open.o): in function `get_attached_info':
hdf5open.c:(.text+0x1e6c): undefined reference to `H5DSget_num_scales'
...
```

**To Reproduce**
```
git clone -b develop --recursive https://github.com/noaa-emc/hpc-stack hpc-stack-cheyenne
cd hpc-stack-cheyenne
./setup_modules.sh -p /glade/work/heinzell/fv3/hpc-stack-cheyenne-gnu-10.1.0 -c config/config_cheyenne_gnu.sh 2>&1 | tee setup_modules.log
./build_stack.sh -p /glade/work/heinzell/fv3/hpc-stack-cheyenne-gnu-10.1.0 -c config/config_cheyenne_gnu.sh -y config/stack_ncar.yaml -m 2>&1 | tee build_stack.try01.log
```
**Expected behavior**
No errors during installation of cdo.

**System:**
Cheyenne with any Intel or GNU compiler

**Comment by aerorahul** [comment]
@kgerheiser Do you have Cheyenne access? I don't so it is not possible for me to debug this particular issue.
I am wondering if this a CDO build issue, since liking to hdf5 errors would be present in other packages that depend on HDF5.

**Comment by kgerheiser** [comment]
No, I don't. I think I've seen this error before, but it seems to be system dependent.

**Comment by edwardhartnett** [comment]
There are two HDF5 libraries, libhdf5 and libhdf5_hl ("high level"). I think you are missing the hl one...

**Comment by aerorahul** [comment]
I don't think so @edwardhartnett :
```
/glade/work/heinzell/fv3/hpc-stack-cheyenne-gnu-10.1.0/gnu-10.1.0/mpt-2.22/hdf5/1.10.6/lib/libhdf5_hl.a

/glade/work/heinzell/fv3/hpc-stack-cheyenne-gnu-10.1.0/gnu-10.1.0/mpt-2.22/hdf5/1.10.6/lib/libhdf5.a 
```

**Comment by edwardhartnett** [comment]
Well OK, I see that you are correct. However, the missing HDF5 functions are in the _hl library, IIRC...

**Comment by climbfuji** [comment]
I recall that sometimes, for unknown reasons and only on some machines, the linker order (first hdf5 then hdf5_hl or vice versa) gets messed up.

**Comment by edwardhartnett** [comment]
Yes, this is explained here:
https://stackoverflow.com/questions/45135/why-does-the-order-in-which-libraries-are-linked-sometimes-cause-errors-in-gcc

However, the order appears correct, first netcdf, then hdf5_hl, then hdf5.

The linker error is happening in netcdf code (files hdf5open.c and hdf5internal.c are part of the netcdf-c library's libhdf5 layer).

**Comment by kgerheiser** [comment]
I get a similar error on WCOSS Dell. Except here it fails when in configuration looking for HDF5. All the flags look correct. Not sure why it's not being picked up.

```
onfigure:24077: checking for library containing H5Fopen
configure:24108: mpiicc -std=gnu99 -o conftest   -fPIC -fopenmp  -I/gpfs/dell2/emc/modeling/noscrub/Kyle.Gerheiser/hpc-stack-cdo/install/ips-18.0.1.163/impi-18.0.1/hdf5/1.10.6/include -DgFortran -L/gpfs/dell2/emc/modeling/noscrub/Kyle.Gerheiser/hpc-stac\
k-cdo/install/ips-18.0.1.163/impi-18.0.1/hdf5/1.10.6/lib -L/gpfs/dell2/emc/modeling/noscrub/Kyle.Gerheiser/hpc-stack-cdo/install/ips-18.0.1.163/impi-18.0.1/netcdf/4.7.4/lib -lnetcdf -L/gpfs/dell2/emc/modeling/noscrub/Kyle.Gerheiser/hpc-stack-cdo/install\
/ips-18.0.1.163/impi-18.0.1/hdf5/1.10.6/lib -lhdf5_hl -lhdf5   -L/gpfs/dell2/emc/modeling/noscrub/Kyle.Gerheiser/hpc-stack-cdo/install/ips-18.0.1.163/zlib/1.2.11/lib  -lz -ldl -lm   conftest.c  -lm  >&5
icc: warning #10315: specifying -lm before files may supersede the Intel(R) math library and affect performance
/tmp/iccF0MhyX.o: In function `main':
conftest.c:(.text+0x3c): undefined reference to `H5Fopen'
```

**Comment by kgerheiser** [comment]
I found the problem for WCOSS. It might be relevant here.

For some reason linker flags that come before the file name when compiling aren't picked up. We set `LDFLAGS` to be `-lhdf5_hl -lhdf5, etc` and then the compile command is `mpiicc <LDFLAGS> conftest.c`, but if you set the `LIBS` variable the command becomes `mpiicc conftest.c <LIBS>`, and it works.

**Comment by kgerheiser** [comment]
I think our usage is prevalent in a lot of libraries and it's not quite right. Pretty sure `LDFLAGS` should just contain `-L<path>` and `LIBS` should be `-l<lib>`.

**Comment by climbfuji** [comment]
> I think our usage is prevalent in a lot of libraries and it's not quite right. Pretty sure `LDFLAGS` should just contain `-L<path>` and `LIBS` should be `-l<lib>`.

I agree, thanks for looking into this.

**Comment by climbfuji** [comment]
@kgerheiser if you are going to fix this in develop, please let me know and I'll give it a try on Cheyenne. Thanks!

**Comment by kgerheiser** [comment]
Fixed with #204
